### PR TITLE
Add frameId from image path & new function for feature loading

### PR DESCRIPTION
**Pull request by gregoire-dl**
## Description

This PR adds required features for the `QtAliceVision` plugin development.

## Features list

- [x] New function in `ViewIO` to detect image sequence from image path
- [x] Try to detect `frameId` from image path in `CameraInit` software
- [x] New function for feature loading in `RegionsIO` (for outside library use)

### CODE REVIEW

File: src/aliceVision/sfm/pipeline/regionsIO.cpp [hunk 1]
```diff
@@ -128,6 +128,58 @@ std::unique_ptr<feature::Regions> loadFeatures(const std::vector<std::string>& f
   return regionsPtr;
 }
 
+bool loadFeaturesPerDescPerView(std::vector<std::vector<std::unique_ptr<feature::Regions>>>& featuresPerDescPerView,
+                                const std::vector<IndexT>& viewIds, const std::vector<std::string>& folders,
+                                const std::vector<feature::EImageDescriberType>& imageDescriberTypes)
+{
+  if(folders.empty())
+  {
+    ALICEVISION_LOG_ERROR("Can't load features, no folders provided");
```
Comment (simogasp) [inline src/aliceVision/sfm/pipeline/regionsIO.cpp:137]:
```suggestion
    ALICEVISION_LOG_ERROR("Cannot load features, no folders provided");
```

File: src/aliceVision/sfm/pipeline/regionsIO.cpp [hunk 2]
```diff
@@ -128,6 +128,58 @@ std::unique_ptr<feature::Regions> loadFeatures(const std::vector<std::string>& f
...
+    return false;
+  }
+  if(viewIds.empty())
+  {
+    ALICEVISION_LOG_ERROR("Can't load features, no view ids provided");
```
Comment (simogasp) [inline src/aliceVision/sfm/pipeline/regionsIO.cpp:142]:
```suggestion
    ALICEVISION_LOG_ERROR("Cannot load features, no view ids provided");
```

File: src/aliceVision/sfm/pipeline/regionsIO.cpp [hunk 3]
```diff
@@ -128,6 +128,58 @@ std::unique_ptr<feature::Regions> loadFeatures(const std::vector<std::string>& f
...
+    return false;
+  }
+  if(imageDescriberTypes.empty())
+  {
+    ALICEVISION_LOG_ERROR("Can't load features, no image desciber types provided");
```
Comment (simogasp) [inline src/aliceVision/sfm/pipeline/regionsIO.cpp:147]:
```suggestion
    ALICEVISION_LOG_ERROR("Cannot load features, no image desciber types provided");
```

File: src/aliceVision/sfmDataIO/viewIO.cpp [hunk 4]
```diff
@@ -279,5 +279,30 @@ std::vector<std::string> viewPathsFromFolders(const sfmData::View& view, const s
...
+    // regexFrame: ^(.*\D)?([0-9]+)([\-_\.].*[[:alpha:]].*)?$
+    std::regex regexFrame("^(.*\\D)?"       // the optional prefix which end with a non digit character
+                          "([0-9]+)"        // the sequence frame number
+                          "([\\-_\\.]"      // the suffix start with a separator
+                          ".*[[:alpha:]].*" // at least one letter in the suffix
+                          ")?$"             // suffix is optional
+    );
+
+    std::smatch matches;
+    if(std::regex_search(imagePathStem, matches, regexFrame))
```
Comment (simogasp) [inline src/aliceVision/sfmDataIO/viewIO.cpp:296]:
```suggestion
    const bool isSequence = std::regex_search(imagePathStem, matches, regexFrame)
    if(isSequence)
```

File: src/aliceVision/sfmDataIO/viewIO.hpp [hunk 5]
```diff
@@ -91,5 +91,16 @@ std::shared_ptr<camera::IntrinsicBase> getViewIntrinsic(
 */
 std::vector<std::string> viewPathsFromFolders(const sfmData::View& view, const std::vector<std::string>& folders);
 
+
+/*
+* @brief Allows to detect if an image is part of an image sequence from the analysis of its filename
```
Comment (simogasp) [inline src/aliceVision/sfmDataIO/viewIO.hpp:96]: You can add here an explanation of the expected pattern

**Review by simogasp — CHANGES_REQUESTED** [review]

File: src/aliceVision/sfmDataIO/viewIO.hpp (hunk 5 above)
Comment (gregoire-dl) [reply src/aliceVision/sfmDataIO/viewIO.hpp:96]: Thanks.
I have made the changes.

File: src/aliceVision/sfmDataIO/viewIO.hpp [hunk 6]
```diff
@@ -93,7 +93,8 @@ std::vector<std::string> viewPathsFromFolders(const sfmData::View& view, const s
 
 
 /*
-* @brief Allows to detect if an image is part of an image sequence from the analysis of its filename
+* @brief Allows to detect if an image is part of an image sequence from the analysis of its filename.
+*        Expected pattern: (optional prefix which end with a non digit character)(the sequence frame number)(optional suffix with at least one letter which start with a separator)
```
Comment (simogasp) [inline src/aliceVision/sfmDataIO/viewIO.hpp:97]:
```suggestion
*        Expected pattern: (optional prefix which end with a non digit character)(the sequence frame number)(optional suffix with at least one letter which start with a separator ('.', '-', '_'))
          Examples:
          IMG01234
          IMG_01234.cam
          C4M0123-A
          01234
```

File: src/software/pipeline/main_cameraInit.cpp [hunk 7]
```diff
@@ -454,6 +454,18 @@ int aliceVision_main(int argc, char **argv)
...
     }
+
+    // try to detect image sequence
+    {
+        IndexT frameId;
+        std::string prefix;
+        std::string suffix;
+        if(sfmDataIO::detectImageSequenceFromImagePath(fs::path(view.getImagePath()).stem().string(), frameId, prefix, suffix))
+        {
+          view.setFrameId(frameId);
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp:465]: There is another setFrameId in the same file that needs to be removed now.

Comment (fabiencastan) [reply src/software/pipeline/main_cameraInit.cpp:465]: It was dedicated for the rig case, but this one should do the job.

File: src/aliceVision/sfmDataIO/viewIO.hpp [hunk 8]
```diff
@@ -91,5 +91,22 @@ std::shared_ptr<camera::IntrinsicBase> getViewIntrinsic(
...
+
+/*
+* @brief Allows to detect if an image is part of an image sequence from the analysis of its filename.
+*        Expected pattern: (optional prefix which end with a non digit character)(the sequence frame number)(optional suffix with at least one letter which start with a separator ('.', '-', '_'))
+          Examples:
+          IMG01234
+          IMG_01234.cam
+          C4M0123-A
+          01234
+* @param[in] imagePathStem The image path
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/viewIO.hpp:103]:
```suggestion
* @param[in] imagePathStem the filename stripped of its extension
```

File: src/aliceVision/sfmDataIO/viewIO.cpp [hunk 9]
```diff
@@ -279,5 +279,29 @@ std::vector<std::string> viewPathsFromFolders(const sfmData::View& view, const s
     });
 }
 
+bool detectImageSequenceFromImagePath(const std::string& imagePathStem, IndexT& frameId, std::string& prefix, std::string& suffix)
```
Comment (fabiencastan) [inline src/aliceVision/sfmDataIO/viewIO.cpp:282]:
```suggestion
bool extractNumberFromFileStem(const std::string& imagePathStem, IndexT& frameId, std::string& prefix, std::string& suffix)
```

File: src/software/pipeline/main_cameraInit.cpp [hunk 10]
```diff
@@ -440,11 +440,9 @@ int aliceVision_main(int argc, char **argv)
     {
       try
       {
-        const int frameId = std::stoi(fs::path(view.getImagePath()).stem().string());
         const int subPoseId = std::stoi(parentPath.stem().string());
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp:443]: Use extractNumberFromFileStem to extract the subPoseId.

File: src/software/pipeline/main_cameraInit.cpp [hunk 11]
```diff
@@ -440,7 +440,14 @@ int aliceVision_main(int argc, char **argv)
     {
       try
       {
-        const int subPoseId = std::stoi(parentPath.stem().string());
+        IndexT subPoseId;
+        std::string prefix;
+        std::string suffix;
+        if(!sfmDataIO::extractNumberFromFileStem(fs::path(view.getImagePath()).stem().string(), subPoseId, prefix, suffix))
+        {
+          throw std::exception("Cannot find sub-pose id from image path.");
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp:448]: ALICEVISION_THROW_ERROR("Cannot find sub-pose id from image path: " << parentPath);

File: src/software/pipeline/main_cameraInit.cpp [hunk 12]
```diff
@@ -440,7 +440,14 @@ int aliceVision_main(int argc, char **argv)
     {
       try
       {
-        const int subPoseId = std::stoi(parentPath.stem().string());
+        IndexT subPoseId;
+        std::string prefix;
+        std::string suffix;
+        if(!sfmDataIO::extractNumberFromFileStem(fs::path(view.getImagePath()).stem().string(), subPoseId, prefix, suffix))
```
Comment (fabiencastan) [inline src/software/pipeline/main_cameraInit.cpp:446]: if(!sfmDataIO::extractNumberFromFileStem(parentPath.stem().string(), subPoseId, prefix, suffix))

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**